plugins/rtmp-services/service-specific/twitch.c

```diff
@@ -105,117 +139,190 @@ static bool load_ingests(const char *json, bool write_file)
+			     const char *log_prefix, const char *file_url)
 {
-	if (os_atomic_load_bool(&ingests_refreshed))
+	if (os_atomic_load_bool(&si->ingests_refreshed))
 		return;
 
-	if (!os_atomic_load_bool(&ingests_refreshing)) {
-		os_atomic_set_bool(&ingests_refreshing, true);
+	if (!os_atomic_load_bool(&si->ingests_refreshing)) {
+		os_atomic_set_bool(&si->ingests_refreshing, true);
 
-		twitch_update_info = update_info_create_single(
-			"[twitch ingest update] ", get_module_name(),
-			"https://ingest.twitch.tv/ingests",
-			twitch_ingest_update, NULL);
+		si->update_info =
+			update_info_create_single(log_prefix, get_module_name(),
+						  file_url, ingest_update, si);
 	}
 
 	/* wait five seconds max when loading ingests for the first time */
-	if (!os_atomic_load_bool(&ingests_loaded)) {
+	if (!os_atomic_load_bool(&si->ingests_loaded)) {
 		for (int i = 0; i < seconds * 100; i++) {
-			if (os_atomic_load_bool(&ingests_refreshed)) {
+			if (os_atomic_load_bool(&si->ingests_refreshed)) {
 				break;
 			}
 			os_sleep_ms(10);
 		}
 	}
 }
 
-void load_twitch_data(void)
+void twitch_ingests_refresh(int seconds)
 {
-	char *twitch_cache = obs_module_config_path("twitch_ingests.json");
+	service_ingests_refresh(&twitch, seconds, "[twitch ingest update] ",
+				"https://ingest.twitch.tv/ingests");
+}
 
-	struct ingest def = {.name = bstrdup("Default"),
-			     .url = bstrdup("rtmp://live.twitch.tv/app")};
+void load_service_data(struct service_ingests *si, const char *cache_filename,
+		       struct ingest *def)
+{
+	char *service_cache = obs_module_config_path(cache_filename);
 
-	pthread_mutex_lock(&mutex);
-	da_push_back(cur_ingests, &def);
-	pthread_mutex_unlock(&mutex);
+	pthread_mutex_lock(&si->mutex);
+	da_push_back(si->cur_ingests, def);
+	pthread_mutex_unlock(&si->mutex);
 
-	if (os_file_exists(twitch_cache)) {
-		char *data = os_quick_read_utf8_file(twitch_cache);
+	if (os_file_exists(service_cache)) {
+		char *data = os_quick_read_utf8_file(service_cache);
 		bool success;
 
-		pthread_mutex_lock(&mutex);
-		success = load_ingests(data, false);
-		pthread_mutex_unlock(&mutex);
+		pthread_mutex_lock(&si->mutex);
+		success = load_ingests(si, data, false);
+		pthread_mutex_unlock(&si->mutex);
 
 		if (success) {
-			os_atomic_set_bool(&ingests_loaded, true);
+			os_atomic_set_bool(&si->ingests_loaded, true);
 		}
 
 		bfree(data);
 	}
 
-	bfree(twitch_cache);
+	bfree(service_cache);
+}
+
+void load_twitch_data(void)
+{
+	struct ingest def = {.name = bstrdup("Default"),
+			     .url = bstrdup("rtmp://live.twitch.tv/app")};
+	load_service_data(&twitch, "twitch_ingests.json", &def);
+}
+
+void unload_service_data(struct service_ingests *si)
+{
+	update_info_destroy(si->update_info);
+	free_ingests(si);
+	pthread_mutex_destroy(&si->mutex);
 }
 
 void unload_twitch_data(void)
 {
-	update_info_destroy(twitch_update_info);
-	free_ingests();
-	pthread_mutex_destroy(&mutex);
+	unload_service_data(&twitch);
+}
+
+void init_amazon_ivs_data(void)
```

Comment:
@PatTheMav Please take a look at the latest commit I just pushed. I've refactored into separate Twitch and Amazon IVS files, and put the common code in `service-ingest.c`.

Reply:
Seems like it addresses my concerns at first glance (mainly making it clear that the relationship is `Twitch IS A Amazon IVS implementation` and not the other way around).

Will take another look at the `service-ingest` file because the file name suggests that it's a fully generic ingest implementation (i.e. could also be used to implement a different ingest service that is used like a white label product).

Reply:
As far as I can tell, the only concern is the generic name. However, off-thread discussion concluded that the concern is not blocking. If this _is_ a blocking concern, please block.